mesonbuild/interpreter.py

```diff
@@ -2189,6 +2196,8 @@ def _get_native() -> str:
                                   'depfile',
                                   'capture',
                                   'preserve_path_from'},
+                    'generator_target': {'output',
+                                         'sources'},
```

Comment:
Here we define the permitted kwargs for this function as `output` and `sources`, but then later we also use `command` and `input`? 

From what I understand, the main gripe people have with `custom_target()` is that you have to define it over and over again even if you just want to call the same transformation on different files. So they try to use `generator()` and trip over its deficiencies: that it's called repeatedly if you use the output of `gen_object.process()` in multiple places and that you can't chain it.

So I was imagining that `generator_target()` would similarly only take `command:` + `output:` kwargs, etc (no `input:`) and return an object with, say, a `run()` method that takes a `input:` kwarg (a list), which would create a target with the specified inputs and outputs. Something like:

```meson
gen = generator_target('gen1',
    command: [find_program('script.py'), '@INPUT@', '@OUTPUT@'],
    output: ['@BASENAME@.c'])
# This will generate common1.c common2.c
common_sources = gen.run(input: ['common1.in', 'common2.in'])
# This will generate source1.c source2.c, and reuse common1.c common2.c
executable('some-tool1', gen.run(input: ['source1.in', 'source2.in']), common_sources)
# This will reuse source1.c, generate source3.c, and reuse common1.c common2.c
executable('some-tool2', gen.run(input: ['source1.in', 'source3.in']), common_sources)
```

Reply:
We will probably want to provide smarter syntax for the `command:` and `output:` templates, since there are a bunch of common cases to be covered. For example, commands that only take one input and one output vs commands that can take multiple inputs and multiple outputs, and commands that can take multiple inputs that are then converted to exactly those many outputs, etc.

Reply:
Alteratively one could do this:

```meson
g = generator(...)
generator_target('foo', 'input1.txt',
  generator: g)
```

The difference to custom target being that the output goes in the output dir rather than current build dir.